src/main/java/BinaryTree/AVLtree.java

```diff
@@ -0,0 +1,344 @@
+
+            // Deleting a node with no children
+            if (current.getLeftNode() == null && current.getRightNode() == null) {
+                if (current.getParentNode().getKey() < current.getKey()) {
+                    current.getParentNode().setRightNode(null);
+                } else {
+                    current.getParentNode().setLeftNode(null);
+                }
+
+            }
+            // Deleting a node with one child
+            else if (current.getLeftNode() == null && current.getRightNode() != null) {
+                if (current.getParentNode() == null) { // Removing a root node
+                    current.getRightNode().setParentNode(null);
+                    current = current.getRightNode();
+                } else {
+                    if (current.getParentNode().getKey() < current.getRightNode().getKey()) {
+                        current.getParentNode().setRightNode(current.getRightNode());
+                        current.getRightNode().setParentNode(current.getParentNode());
+                    } else {
+                        current.getParentNode().setLeftNode(current.getRightNode());
+                        current.getRightNode().setParentNode(current.getParentNode());
+                    }
+                }
+            } else if (current.getLeftNode() != null && current.getRightNode() == null) {
+                if (current.getParentNode() == null) { // removing a root node
+                    current.getLeftNode().setParentNode(null);
+                    current = current.getLeftNode();
+                } else {
+                    if (current.getParentNode().getKey() < current.getRightNode().getKey()) {
+                        current.getParentNode().setRightNode(current.getLeftNode());
+                        current.getLeftNode().setParentNode(current.getParentNode());
+                    } else {
+                        current.getParentNode().setLeftNode(current.getLeftNode());
+                        current.getLeftNode().setParentNode(current.getParentNode());
+                    }
+                }
+            }
+            // Deleting a node with two children
+            else {
+                BinaryNode<T> successor = find_successor();
+                bst_delete(successor.getKey());
+                current.setKey(successor.getKey());
+            }
+        }
+        updateHeight(current);
+    }
+
+    public BinaryNode<T> find_successor() {
+        BinaryNode<T> original_position = current;
+        moveToRightNode();
+        while (current.getLeftNode() != null) {
+            moveToLeftNode();
+        }
+        BinaryNode<T> successor = current;
+        current = original_position;
+        return successor;
+    }
+
+    public void delete(int delete_key) {
+        // perform standard bst deletion
+        bst_delete(delete_key);
+
+        // balance the tree
+        int depth_difference = getDepthDifference(current);
```

Comment:
create a function `balance_tree_depth` and call it from both insert and delete

Reply:
insert and deletion have different balancing conditions